### DISCUSSION TITLE
Two approaches, different results for SFC WIND PR_CORR

### DISCUSSION THREAD

**Discussion by legibe** [dtcenter/METplus#1972] [Statistical Computation] [answered]
Hello,

In order to speed up the calculation of scores, I decided to skip a step that is not necessary for a particular use case. We are verifying surface forecasts against metar. The usual steps we take are as follows:

- execute **point_stat** forecast metar for each cycle
- execute **stat_analysis** on the output files: job=aggregate_stat, -line_type=MPR,  SL1L2, -out_line_type SL1L2, -by=OBS_SID,FCST_LEAD,FCST_VAR,FCST_INIT_HOUR
- execute **stat_analysis** on the output files of the previous stat_analysis: job=aggregate_stat, -line_type=SL1L2,  SL1L2, -out_line_type CNT, -by=FCST_LEAD,FCST_VAR,FCST_INIT_HOUR,FCST_INIT_BEG

The second approach has two steps:
- execute **point_stat** forecast metar for each cycle
- execute **stat_analysis** on the output files: job=aggregate_stat, -line_type=SL1L2,  SL1L2, -out_line_type CNT, -by=FCST_LEAD,FCST_VAR,FCST_INIT_HOUR

The second approach is much faster to execute and produce 99.9% identical results. There is a special case for the score PR_CORR for the WIND variable for which, in some instances, we get NA using method one and we get a very small values instead with method 2: eg. ~10e-7. The second approach does also produce some NA values in that use case, but for some reason the results are not identical. Both approached use the same configuration file(s).

Here is a small example:

Approach 1:
ts: 000000 WIND PR_CORR sfc 0.87041, 0.8522, 0.78686, 0.8572, 0.83264, **nan**, 0.81251, 0.8472, 0.74564, 0.79858, 0.66196, 0.61759, 0.66879, 0.73828, nan, 0.79355, 0.77764, 0.79359, nan, 0.8971
ts: 210000 WIND PR_CORR sfc 0.79418, nan, 0.78741, 0.74796, 0.75412, nan, 0.78185, 0.50505, 0.61448, 0.49433, 0.73533, 0.65687, 0.69892, nan, 0.83441, 0.88799, nan, 0.78103, 0.70363, **nan**

Approach 2:
ts: 000000 WIND PR_CORR sfc 0.87041, 0.8522, 0.78686, 0.8572, 0.83264, **-2.0775e-08,** 0.81251, 0.8472, 0.74564, 0.79858, 0.66196, 0.61759, 0.66879, 0.73828, 0.0, 0.79355, 0.77764, 0.79359, nan, 0.8971
ts: 210000 WIND PR_CORR sfc 0.79418, nan, 0.78741, 0.74796, 0.75412, -2.2362e-08, 0.78185, 0.50505, 0.61448, 0.49433, 0.73533, 0.65687, 0.69892, nan, 0.83441, 0.88799, nan, 0.78103, 0.70363, **2.6601e-07**

I would have expected both approaches to produce identical results. I thought I would let you know about this feature. This is not a showstopper for us, but I would rather let MetPlus make the decisions about what should be NA or not.

**Comment by JohnHalleyGotway** [comment]
@legibe thanks for bringing this to our attention. And thanks for laying out the 2 different approaches. Yes, I definitely see why approach 2 is much, much faster than approach 1. Generally speaking, we provided the individual matched pairs in the MPR line type for debugging and digging into details. We did not intend them to be included in the logic for processing bulk statistics.

It certainly is reasonable to list out a handful of individual stations for which you want matched pairs, but writing/processing them for ALL stations will be slow. So if you can, writing/aggregating partial sums is the preferred approach.

But I'd say you've actually uncovered a bug in our library code. Seeing **nan** in the output is ALWAYS bad. Seeing **NA** is fine, but **nan** indicates that we're missing some error checking along the way and are likely taking the square root of a negative number or dividing by 0 somewhere.

Can you please send me some sample data to demonstrate the creation of these **nan** values? Presumably, I'd just need a the MPR and SL1L2 data that's written by Point-Stat and passed as input to Stat-Analysis. I'd like to replicate these **nan** values and then patch up the code for the next officially release.

We're actually working toward a MET 11.0.0 release in the next 2 weeks. If you're able to send me sample data soon, I should be able to include a fix in that upcoming version.

Thanks!

**Comment by legibe** [comment]
Hello John,

many thanks for your swift reply. Perhaps I wasn't clear, the examples I showed are produced by a test program that reads the results in Python, wherever it said **nan** (numpy.nan) it meant that the value NA was found in the MetPlus ASCII, CNT line file.I think the error is when approach 1 generates a NA value and for the same instance approach 2 generated a 10e-7 value.
I am very new to MetPlus and I am not fully sure about creating a small working example of this. What you would need exactly?

**Comment by legibe** [comment, marked as answer]
John,

you wrote: "_we provided the individual matched pairs in the MPR line type for debugging and digging into details_. we provided the individual matched pairs in the MPR line type for debugging and digging into details."
Is there another way of doing this?

**Reply by JohnHalleyGotway** [reply]
@legibe, ah OK, I'm much less concerned if those *nan*'s are being written by python, and not the MET tools themselves.

I am not actually all that concerned about getting NA from one method and something very close to 0 from another. I suspect that if we dig into the details we'd end up chalking this up to machine precision. The value of 1.0e-8 is extremely close to the value of -1.0e-8. The square root of the first yields a 1.0e-4 while the square root of the second yields bad data, because you can't compute the square root of a negative number.

The precision of data is an ongoing issue, especially for numbers very close to 0. Many versions ago, we started using the [is_eq()](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/src/basic/vx_math/is_bad_data.h#L54) function in MET. Instead of checking for exact equality of double precision numbers, we check to see if the absolute value of their difference is less than a tolerance (tight_tol = 10e-10, by default). And in some cases we only require them to be within 10e-5 (loose_tol = 10e-5).

We could potentially address this by writing a wrapper function for square roots to first check if a number is very, very slightly negative and, if so, set it to 0. I'm not positive that that's what going on here, but I suspect it's something like this.

I wouldn't say that either answer you're seeing is obviously incorrect, but I do understand its uncomfortable that they're not the same.

If you'd like me to look more closely at this, I'd just need the `.stat` files created by your calls to Point-Stat.

**Reply by JohnHalleyGotway** [reply]
Hi @legibe, wondering you have any additional questions about this discussion? If not, I'll mark it as being answered and lock it to prevent future posts. That makes it easier for us to keep track of active discussions. Of course, if new issues or questions arise, you can always create a new discussion.

**Comment by legibe** [comment]
Hello John,

many thanks, you can close it, I have made good progress.

Claude
